# Introductory data analysis with a real world dataset

In this notebook, we'll cover introductory Data Analysis skills that can help you analyze new datasets in real life. We've been exploring different pandas techniques in the previous problem sets. Let's see how we can apply those techniques to clean, explore, and gain insights from the "Most Watched Netflix Original Shows" dataset. Credit: [Muhmad Akmal,  Kaggle](https://www.kaggle.com/code/muhmadakmal/most-watched-netflix-original-shows-tv-time/input)


## Loading the data

Let's load the data in pandas and glance at the first few rows.

In [33]:
import pandas as pd
df = pd.read_csv('netflix_data.csv')

Use `df.head()` to return the first n rows of a dataframe based on position. By default,`df.head()` will return the first 5 rows 


In [2]:
df.head()

lister-item-index       lister-item-header certificate runtime  \
0                  1          Stranger Things          15  60 min   
1                  2           13 Reasons Why          18  60 min   
2                  3  Orange Is the New Black          18  59 min   
3                  4             Black Mirror          18  60 min   
4                  5              Money Heist          15  60 min   

                      genre  rating      votes  
0    Drama, Fantasy, Horror     8.7  1,327,188  
1  Drama, Mystery, Thriller     7.5    314,321  
2      Comedy, Crime, Drama     8.0    319,342  
3    Drama, Mystery, Sci-Fi     8.7    636,319  
4      Action, Crime, Drama     8.2    529,086

Let's say we are interested in seeing the first 8 rows. We can simply specify:

In [3]:
df.head(8)

lister-item-index       lister-item-header certificate runtime  \
0                  1          Stranger Things          15  60 min   
1                  2           13 Reasons Why          18  60 min   
2                  3  Orange Is the New Black          18  59 min   
3                  4             Black Mirror          18  60 min   
4                  5              Money Heist          15  60 min   
5                  6                  Lucifer          15  60 min   
6                  7                   Narcos          15  50 min   
7                  8                Daredevil          15  60 min   

                      genre  rating      votes  
0    Drama, Fantasy, Horror     8.7  1,327,188  
1  Drama, Mystery, Thriller     7.5    314,321  
2      Comedy, Crime, Drama     8.0    319,342  
3    Drama, Mystery, Sci-Fi     8.7    636,319  
4      Action, Crime, Drama     8.2    529,086  
5     Crime, Drama, Fantasy     8.1    354,155  
6   Biography, Crime, Drama     8.8    467,909  
7      Action, Crime, Drama     8.6    472,940

`df.tail()` works in a similar way. It returns the last n rows of the dataset, with the default being 5 rows. 

In [4]:
df.tail()

lister-item-index lister-item-header certificate runtime  \
75                 76    F Is for Family          15  30 min   
76                 77          The Ranch          15  30 min   
77                 78    American Vandal          15  34 min   
78                 79         Dead to Me          15  30 min   
79                 80          Quicksand          18  45 min   

                       genre  rating   votes  
75  Animation, Comedy, Drama     8.0  41,074  
76    Comedy, Drama, Western     7.5  42,401  
77      Comedy, Crime, Drama     8.1  32,985  
78      Comedy, Crime, Drama     7.9  99,440  
79     Crime, Drama, Mystery     7.4  25,507

You can use `df.sample()` to take a peek at random rows in the dataset. This is especially useful in cases where you don't think the first/last rows are representative of the entire dataset.

In [5]:
df.sample(5)

lister-item-index         lister-item-header certificate runtime  \
40                 41  Unbreakable Kimmy Schmidt          12  30 min   
42                 43                Cable Girls          15  50 min   
52                 53       Love, Death & Robots          18  15 min   
64                 65          One Day at a Time          15  30 min   
31                 32             Altered Carbon          18  60 min   

                       genre  rating    votes  
40                    Comedy     7.6   78,450  
42                     Drama     7.5   16,087  
52  Animation, Short, Action     8.4  195,422  
64             Comedy, Drama     8.2   18,238  
31     Action, Drama, Sci-Fi     7.9  187,859

I'll also briefly explain what each column represents:

`lister-item-index:` Describes the index of the show as listed on IMDB, and begins indexing at 1.

`lister-item-header:` Show Name.

`certificate:` Maturity Rating. For example, a certificate of 15 implies that content may be unsuitable for children under the age of 15. 

`runtime:` The average runtime of each episode.

`genre:` The genres associated with the show. 

`rating:` IMDB rating.

`votes:` Number of people that rated the show.

For the purposes of this exercise, let's assume that all the data in this dataset is accurate. 

## Analyzing Columns & Rows

Now that we have a good idea of what this dataset looks like, let's analyze the columns more closely! 

In this case, we can visually see the title of all 7 columns of our dataset. But in larger datasets (say with a 100 columns) we can use `df.columns` to return all column labels.


In [6]:
df.columns

Index(['lister-item-index', 'lister-item-header', 'certificate', 'runtime',
       'genre', 'rating', 'votes'],
      dtype='object')

Some of these column names are little long. Luckily, we can use `df.rename` to rename any column we'd like. 

In [7]:
#this code generates a *new dataframe* with the renamed columns from df
df.rename(columns={'lister-item-header': 'show', 'lister-item-index': 'index'})

index                     show certificate runtime  \
0       1          Stranger Things          15  60 min   
1       2           13 Reasons Why          18  60 min   
2       3  Orange Is the New Black          18  59 min   
3       4             Black Mirror          18  60 min   
4       5              Money Heist          15  60 min   
..    ...                      ...         ...     ...   
75     76          F Is for Family          15  30 min   
76     77                The Ranch          15  30 min   
77     78          American Vandal          15  34 min   
78     79               Dead to Me          15  30 min   
79     80                Quicksand          18  45 min   

                       genre  rating      votes  
0     Drama, Fantasy, Horror     8.7  1,327,188  
1   Drama, Mystery, Thriller     7.5    314,321  
2       Comedy, Crime, Drama     8.0    319,342  
3     Drama, Mystery, Sci-Fi     8.7    636,319  
4       Action, Crime, Drama     8.2    529,086  
..                       ...     ...        ...  
75  Animation, Comedy, Drama     8.0     41,074  
76    Comedy, Drama, Western     7.5     42,401  
77      Comedy, Crime, Drama     8.1     32,985  
78      Comedy, Crime, Drama     7.9     99,440  
79     Crime, Drama, Mystery     7.4     25,507  

[80 rows x 7 columns]

To keep track of all the changes we're making (like renaming columns), we'll  maintain two version of our dataset! `df` will contain the original dataframe we loaded from the `pd.read_csv` call above. 

Meanwhile, we'll use `new_df` to store the cleaned version of our data. At the end, we'll compare `df` and `new_df`so you can see all the differences.

In [8]:
#this codes *assigns* the result of the renaming operation above to new_df
new_df = df.rename(columns={'lister-item-header': 'show', 'lister-item-index': 'index'})

You can see that `df` remains the same.

In [9]:
df.head(1)

lister-item-index lister-item-header certificate runtime  \
0                  1    Stranger Things          15  60 min   

                    genre  rating      votes  
0  Drama, Fantasy, Horror     8.7  1,327,188

And that the new column names are only reflected in `new_df`. 

In [10]:
new_df.head(1)

index             show certificate runtime                   genre  rating  \
0      1  Stranger Things          15  60 min  Drama, Fantasy, Horror     8.7   

       votes  
0  1,327,188

Let's keep going with our analysis. We can use `df.shape` to return the number of rows and columns in our orginal dataframe. 

In [11]:
df.shape

(80, 7)

Note that `df.shape` returns a tuple, which we can actually store as two seperate variables for easy future access.

In [12]:
rows, columns = df.shape

In [13]:
rows

80

In [14]:
columns

7

There are some other helpful things we can use to describe our data! 

`df.info()` prints useful information about the dataframe, including the data types found in each column.

In [15]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 80 entries, 0 to 79
Data columns (total 7 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   lister-item-index   80 non-null     int64  
 1   lister-item-header  80 non-null     object 
 2   certificate         80 non-null     object 
 3   runtime             80 non-null     object 
 4   genre               80 non-null     object 
 5   rating              80 non-null     float64
 6   votes               80 non-null     object 
dtypes: float64(1), int64(1), object(5)
memory usage: 4.5+ KB


The `Non-Null Count` tells us the number of non null values in that column.

`Dtype` tells us the data type of the values in a column: 
* lister-item-index is of type `int64`, 
* rating is of type `float64`
* the remaining  columns are of type `object`.

In other words, only two of our columns contain numerical values (lister-item-index and rating). The lister-item-header, certificate, genre, and votes columns are all strings (the default data type for strings in DataFrames is `object`).


We also get the overall memory usage of our dataframe. Side note: If you're interested in a more accurate measure of memory usage, feel free to try `df.info(memory_usage='deep')`

Small segway since we're talking about null values! There are many, many, many ways to check for null values in your dataset! Here's another approach:

In [16]:
df.isnull().values.any()

False

You could also run the following code to see _how many_ null values you have across the entire dataset. In our case, we know the total number is 0.


In [17]:
df.isnull().sum().sum()

0

Another useful method (and one of my favorites) is `df.describe()`. `df.describe()` generates descriptive statistics (such as total count, mean, standard deviation, min, and max) for all _numerical_ columns in your dataset by default. 

As we saw above, only the `lister-item-index` and `rating` columns in our dataframe are numerical.

In [18]:
df.describe()

lister-item-index    rating
count            80.0000  80.00000
mean             40.5000   7.72000
std              23.2379   0.71235
min               1.0000   5.90000
25%              20.7500   7.20000
50%              40.5000   7.80000
75%              60.2500   8.22500
max              80.0000   8.80000

It's not particularly helpful to learn statistics about the `lister-item-index`	column- the index just identifies each row in the dataset. 


_But_ we did get some interesting analysis on the ratings across Netflix Orginal shows! For example, the average rating for a show was a 7.72. The standard deviation was only around .71. We also learned that the max rating achieved by any show was 8.8

If you'd like, you could also specify your own percentiles for `df.describe().`

In [19]:
percentiles = [.05, .10, .25, .50, .75, .90, .99]
df.describe(percentiles)

lister-item-index    rating
count            80.0000  80.00000
mean             40.5000   7.72000
std              23.2379   0.71235
min               1.0000   5.90000
5%                4.9500   6.40000
10%               8.9000   6.70000
25%              20.7500   7.20000
50%              40.5000   7.80000
75%              60.2500   8.22500
90%              72.1000   8.60000
99%              79.2100   8.80000
max              80.0000   8.80000

Let's come back to `df.describe()` after we do some more data cleaning. This method will become even more useful for us later.

## Data Cleaning

Anubha is a Netflix addict- she'd like to watch the first episode of ALL the Netflix shows in our dataframe. How many minutes will it take her to watch all of these episodes? 

Let's sum up all the runtimes in our table to get the answer. 



In [20]:
df['runtime'].sum()

'60 min60 min59 min60 min60 min60 min50 min60 min60 min56 min50 min25 min60 min60 min60 min45 min60 min60 min60 min60 min55 min55 min30 min30 min45 min50 min60 min60 min45 min60 min49 min60 min60 min30 min30 min25 min26 min60 min60 min50 min30 min30 min50 min60 min58 min60 min60 min30 min23 min30 min60 min30 min15 min30 min60 min25 min63 min60 min30 min45 min30 min50 min60 min27 min30 min60 min35 min60 min51 min52 min45 min30 min22 min46 min60 min30 min30 min34 min30 min45 min'

What went wrong here? We've done this before in previous problem sets! 

Remember that Python only lets us do arithmetic operations on numbers. From `df.info()`, we learned that only `lister-item-index` and `rating` are numerical columns. 

We can also directly see which columns in our dataframe can be summed by running `df.sum(numeric_only= True)`. Runtime is not one of them. 
 

In [21]:
df.sum(numeric_only= True)

lister-item-index    3240.0
rating                617.6
dtype: float64

To sum runtimes, we'll need to convert all values in the runtime column into numerical values instead of a strings (we want to go from _'60 min'_ to  _60_).

Let's do this by first removing the word 'min' from each value, and then converting the result from a string to an int! One approach is to use `df.replace` and `.astype()`


In [22]:
#First, remove the word 'min' from runtimes by replacing it with an empty substring
replace_mins = df['runtime'].str.replace(' min', '')

#Second, convert runtimes to ints 
runtime_as_ints = replace_mins.astype('int')
runtime_as_ints

#You could also do this all in one line by running the following:
#runtime_as_ints = df['runtime'].str.replace(' min', '').astype('int')

0     60
1     60
2     59
3     60
4     60
      ..
75    30
76    30
77    34
78    30
79    45
Name: runtime, Length: 80, dtype: int64

In [23]:
#To actually assign runtime_as_ints to our new_df dataframe, run:
new_df['runtime'] =  runtime_as_ints

In [24]:
#Great! If you run this cell, you'll see that all our runtime values are of type int64!
new_df['runtime']

0     60
1     60
2     59
3     60
4     60
      ..
75    30
76    30
77    34
78    30
79    45
Name: runtime, Length: 80, dtype: int64

Now... how many minutes will it take Anubha to watch the first episode of every show in our Netflix dataset?

In [25]:
new_df['runtime'].sum()

3741

That's a lot of minutes 😬. This example was to show that you'll have to do some data cleaning/data manipulation with pretty much every dataset you encounter. This is a relatively "easy" dataset- it has a usability rating of 10 on Kaggle. But we still had to do some work to find right answer.

Now that we've converted the runtime column into a numerical column, we can do all kinds of other fun analysis. What's the average runtime of Netflix Original shows? 

In [26]:
new_df['runtime'].mean()

46.7625

What about the max runtime?

In [27]:
new_df['runtime'].max()

63

Also, let's go back to `.describe()` and see how the output changes now that runtime is a numerical column! 

In [28]:
new_df.describe()

index    runtime    rating
count  80.0000  80.000000  80.00000
mean   40.5000  46.762500   7.72000
std    23.2379  14.156667   0.71235
min     1.0000  15.000000   5.90000
25%    20.7500  30.000000   7.20000
50%    40.5000  50.000000   7.80000
75%    60.2500  60.000000   8.22500
max    80.0000  63.000000   8.80000

The mode (or most common value) is not included in `.describe()`. Let's see what the most _common_ rating is for a Netflix Orginal show. 



In [29]:
new_df['rating'].mode()

0    8.2
Name: rating, dtype: float64

Note that in our previous mean, max, mode calculations, we didn't actually learn _which_ show is associate with any of those values. Let's compute the max rating again, but this time, let's find out _which_ show got the max rating.


In [30]:
max_rating = new_df['rating'].max()

new_df.loc[new_df['rating'] == max_rating]

index             show certificate  runtime                     genre  \
6       7           Narcos          15       50   Biography, Crime, Drama   
35     36  BoJack Horseman          18       25  Animation, Comedy, Drama   

    rating    votes  
6      8.8  467,909  
35     8.8  186,993

# Exercises 

### 1. Look up your favorite Netflix Original and return its information. If it doesn't pop up in this table, feel free to pick a random show instead.

### 2. What show had the lowest ratings? What was the rating? 

### 3. Which Netflix Original is the most popular (had the most votes)? What was the number of votes?
Hint: Votes are strings here, and need to be converted to numerical values (ints)! 

Another Hint: Last time, we had to remove the word "mins" from runtime values before converting to ints. What do we need to remove from the votes values before we can convert to ints?  


In [31]:
# Make sure to store your new votes values in new_df 

### 4. Sarah loves the Mystery genre. She's curious which shows with the word 'Mystery' in the genre column have the highest ratings. Recommend the top 3 highest rated shows in the Mystery genre to Sarah.



### 5. Leif would like to recommend shows to his 15 year old cousin, but wants to make sure that he recommends shows that are age appropriate. Return a dataframe of all the shows with a Maturity Rating less than or equal to 15.  

Hint: How should we think about handling 'PG' in this case? We could replace 'PG' with 13. 


In [32]:
# Make sure to store your new certificates values in new_df

### 6. Compare `df.info()` and `new_df.info()`. Compare  `df.describe()` and `new_df.describe()`. You can use new_df to answer other questions you are curious about!